dali/pipeline/operators/reader/loader/coco_loader.cc

```diff
@@ -12,433 +12,347 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "dali/pipeline/operators/reader/loader/coco_loader.h"
-
-#include <rapidjson/reader.h>
-#include <rapidjson/document.h>
-
 #include <map>
 #include <unordered_map>
+#include <iomanip>
 
-RAPIDJSON_DIAG_PUSH
-#ifdef __GNUC__
-RAPIDJSON_DIAG_OFF(effc++)
-#endif
-
-using rapidjson::SizeType;
-using rapidjson::Value;
-using rapidjson::Reader;
-using rapidjson::InsituStringStream;
-using rapidjson::kParseDefaultFlags;
-using rapidjson::kParseInsituFlag;
-using rapidjson::kArrayType;
-using rapidjson::kObjectType;
-
-namespace {
-
-// taken from https://github.com/Tencent/rapidjson/blob/master/example/lookaheadparser/lookaheadparser.cpp
-
-class LookaheadParserHandler {
- public:
-  bool Null() { st_ = kHasNull; v_.SetNull(); return true; }
-  bool Bool(bool b) { st_ = kHasBool; v_.SetBool(b); return true; }
-  bool Int(int i) { st_ = kHasNumber; v_.SetInt(i); return true; }
-  bool Uint(unsigned u) { st_ = kHasNumber; v_.SetUint(u); return true; }
-  bool Int64(int64_t i) { st_ = kHasNumber; v_.SetInt64(i); return true; }
-  bool Uint64(uint64_t u) { st_ = kHasNumber; v_.SetUint64(u); return true; }
-  bool Double(double d) { st_ = kHasNumber; v_.SetDouble(d); return true; }
-  bool RawNumber(const char*, SizeType, bool) { return false; }
-  bool String(const char* str, SizeType length, bool) {
-    st_ = kHasString;
-    v_.SetString(str, length);
-    return true;
-  }
-  bool StartObject() { st_ = kEnteringObject; return true; }
-  bool Key(const char* str, SizeType length, bool) {
-    st_ = kHasKey;
-    v_.SetString(str, length);
-    return true;
-  }
-  bool EndObject(SizeType) { st_ = kExitingObject; return true; }
-  bool StartArray() { st_ = kEnteringArray; return true; }
-  bool EndArray(SizeType) { st_ = kExitingArray; return true; }
-
- protected:
-  explicit LookaheadParserHandler(char* str);
-  void ParseNext();
-
- protected:
-  enum LookaheadParsingState {
-    kInit,
-    kError,
-    kHasNull,
-    kHasBool,
-    kHasNumber,
-    kHasString,
-    kHasKey,
-    kEnteringObject,
-    kExitingObject,
-    kEnteringArray,
-    kExitingArray
-  };
-
-  Value v_;
-  LookaheadParsingState st_;
-  Reader r_;
-  InsituStringStream ss_;
-
-  static const int parseFlags = kParseDefaultFlags | kParseInsituFlag;
+#include "dali/pipeline/operators/reader/loader/coco_loader.h"
+#include "dali/pipeline/util/lookahead_parser.h"
+
+namespace dali {
+namespace detail {
+struct ImageInfo {
+  std::string filename_;
+  int original_id_;
+  int width_;
+  int height_;
 };
 
-LookaheadParserHandler::LookaheadParserHandler(char* str) :
-    v_(), st_(kInit), r_(), ss_(str) {
-  r_.IterativeParseInit();
-  ParseNext();
-}
+struct Annotation {
+  int image_id_;
+  int category_id_;
+  std::array<float, 4> box_;
```

Comment:
Could `dali/core/geom/box.h` be useful here?

Reply:
In `COCOReader` depending on the parameters we have different box representations and `dali/core/geom/box.h` is assuming two points, so it might cause confusion, I think.